### DISCUSSION TITLE
Migration from bitwarden

### DISCUSSION THREAD

**Discussion by HikaruIchijyo** [padloc/padloc#551] [Q&A] [answered]
Can you guys allow something like migrating from bitwarden.

Like their data is in csv & json format importing it to padlock will be an awesome addition ig

I literally needed that cuz i have like 200+ pass stores problem is adding 1 - 1 is pain

**Comment by HikaruIchijyo** [comment]
Also second question:

Sorry i am new to app.

But is it possible to add categories or folders support like warden.

Where like u have github folder in which ur all data related to accounts get stored & it is much more organised etc.

Upon searching it can filter folders etc?

**Comment by MaKleSoft** [comment]
Hi @HikaruIchijyo! Thanks for reaching out! We don't have a dedicated import feature for Bitwarden yet (although it's on the roadmap: https://github.com/padloc/padloc/issues/347), but we do have a pretty powerful [CSV import tool](https://docs.padloc.app/manual/settings/#csv-import-options)! Can you give that a shot and let us know if that works for you?

As for folders, we are currently not planning to implement those as we believe that our [tags](https://docs.padloc.app/manual/vaults/#tags) are in fact a superior alternative. It's not *completely* off the table yet though. The feature request is being tracked here: https://github.com/padloc/padloc/issues/533

**Reply by HikaruIchijyo** [reply]
> Hi @HikaruIchijyo! Thanks for reaching out! We don't have a dedicated import feature for Bitwarden yet (although it's on the roadmap: https://github.com/padloc/padloc/issues/347), but we do have a pretty powerful [CSV import tool](https://docs.padloc.app/manual/settings/#csv-import-options)! Can you give that a shot and let us know if that works for you?

I tried it is good i mean importing does work somehow but format is kinda weird after importing i guess i will w8 fr dedicated bitwarden import then.


2nd: if there will be no support for folders try to convert folder data from backups such as from bitwarden & convert them to tags?

**Reply by HikaruIchijyo** [reply]
> Hi @HikaruIchijyo! Thanks for reaching out! We don't have a dedicated import feature for Bitwarden yet (although it's on the roadmap: https://github.com/padloc/padloc/issues/347), but we do have a pretty powerful [CSV import tool](https://docs.padloc.app/manual/settings/#csv-import-options)! Can you give that a shot and let us know if that works for you?
> 
> As for folders, we are currently not planning to implement those as we believe that our [tags](https://docs.padloc.app/manual/vaults/#tags) are in fact a superior alternative. It's not *completely* off the table yet though. The feature request is being tracked here: https://github.com/padloc/padloc/issues/533
> 

Thanks alot for the quick reply😍

**Reply by MaKleSoft** [reply]
> 2nd: if there will be no support for folders try to convert folder data from backups such as from bitwarden & convert them to tags?

You *should* be able to simply map bitwardens folder names to tags in Padloc, although I haven't tried it yet so I'm not sure how well that works in practice. One caveat would certainly be that you'll lose any hierarchical folder structure.

**Reply by HikaruIchijyo** [reply]
> > 2nd: if there will be no support for folders try to convert folder data from backups such as from bitwarden & convert them to tags?
> 
> You *should* be able to simply map bitwardens folder names to tags in Padloc, although I haven't tried it yet so I'm not sure how well that works in practice. One caveat would certainly be that you'll lose any hierarchical folder structure.

I did it kinda messed up data gets saved but format rip

**Reply by doublehelix** [reply]
Have to admit, I tried about 15 times using the CSV mapping functionality to get BitWarden's entries into Padloc. As powerful as this feature seems, I failed every time.
There was always something that didn't work. Sometimes the new entries didn't even show up as website passwords.
Additionally, there are secure notes, two-factor codes etc. and changing built-in field mappings caused the entire mapping process to lose field references and I had to start again. It's a bit complicated.
I'd love to have some (coding) input on this feature when I get time - definitely worth it as I have 1900+ BitWarden entries (I migrated from LastPass)

**Reply by MaKleSoft** [reply]
@doublehelix The CSV import is powerful in the sense that it gives you a lot of options with regards to which columns should be mapped to categories, field types etc, but unfortunately the CSV format itself is quite limited and it's possible that it simply cannot transport all the information you're trying to transfer. Bitwarden is definitely popular enough to warrant a dedicated import feature though, so if you want to take a shot at implementing that, go for it! Their (unencrypted) json format might be a better format for that, though.

**Reply by MaKleSoft** [reply]
Created an issue for it: https://github.com/padloc/padloc/issues/561

**Reply by HikaruIchijyo** [reply]
> Created an issue for it: https://github.com/padloc/padloc/issues/561

also will we have autofill for android?

**Reply by MaKleSoft** [reply]
> also will we have autofill for android?

It is planned, yes: https://github.com/padloc/padloc/issues/250. However we're a small team and currently don't have anyone with enough Android knowledge to knock this out quickly. Both @BrunoBernardino and I would probably be able to get it done given enough time but we've got our hands full with other things right now. So yes, it will definitely be added at some point, but unfortunately I can't give you an ETA at this point.

**Reply by HikaruIchijyo** [reply]
> > also will we have autofill for android?
> 
> It is planned, yes: https://github.com/padloc/padloc/issues/250. However we're a small team and currently don't have anyone with enough Android knowledge to knock this out quickly. Both @BrunoBernardino and I would probably be able to get it done given enough time but we've got our hands full with other things right now. So yes, it will definitely be added at some point, but unfortunately I can't give you an ETA at this point.

got it np take ur time

**Comment by HikaruIchijyo** [comment]
somehow mobile app is still not accepting file(showed blacked cant select file)

while on webapp it shows 0 items upon importing so cant import? i backed up using bitwarden mobile though(nd it does have data)

**Reply by MaKleSoft** [reply]
We only just submitted the mobile apps for review, so you're probably still working with 4.0 which doesn't support bitwarden files yet. It should work find in the web app though! Did you use the unencrypted json export as described here? https://bitwarden.com/help/export-your-data/

**Reply by HikaruIchijyo** [reply]
> We only just submitted the mobile apps for review, so you're probably still working with 4.0 which doesn't support bitwarden files yet. It should work find in the web app though! Did you use the unencrypted json export as described here? https://bitwarden.com/help/export-your-data/

yep & on play store it is already available i mean v4.1.0
yep tried on web & it is unencrypted .json cuz i opened the file locally & it did show

**Reply by BrunoBernardino** [reply]
@HikaruIchijyo can you show us a screenshot of the import modal? Does it identify the selected file as Bitwarden JSON?

**Reply by HikaruIchijyo** [reply]
**web app:**
![Screenshot_2022-09-09-23-15-40-532_com.google.android.documentsui.jpg](https://user-images.githubusercontent.com/38470944/189417663-202f2c36-4d9b-4c69-9ea1-afaa9c1369bd.jpg)
![Screenshot_2022-09-09-23-15-44-238_org.mozilla.fenix.jpg](https://user-images.githubusercontent.com/38470944/189417685-8732269e-ee00-47af-918a-c7b175c80a29.jpg)

simple filed open from phone i cropped username etc.
![IMG_20220909_231647.jpg](https://user-images.githubusercontent.com/38470944/189417759-09908b86-9dee-4727-8cef-bccccf93b512.jpg)

**mobile app:**
here it doesnt even try to select the file
![Screenshot_2022-09-09-23-18-39-215_com.google.android.documentsui.jpg](https://user-images.githubusercontent.com/38470944/189417894-bc756f5f-c987-4c58-be1b-059881fa11ac.jpg)

**Reply by HikaruIchijyo** [reply]
lemme know if u need anything else specifically
or i am doing something wrong i am not sure

**Reply by BrunoBernardino** [reply]
The file format looks good from what can be seen. Is there any error in the browser console?

**Reply by HikaruIchijyo** [reply]
> The file format looks good from what can be seen. Is there any error in the browser console?

nope nothing

**Reply by BrunoBernardino** [reply]
Can you share the file with only 1 item, no _real_ username or password (change to fake values), so I can investigate further? I can't see how to help any other way, I'm afraid.

**Reply by HikaruIchijyo** [reply]
> Can you share the file with only 1 item, no _real_ username or password (change to fake values), so I can investigate further? I can't see how to help any other way, I'm afraid.

well lemme try it first then i can

**Reply by HikaruIchijyo** [reply]
> Can you share the file with only 1 item, no _real_ username or password (change to fake values), so I can investigate further? I can't see how to help any other way, I'm afraid.

https://file.io/rMuKRpGt73wf
about web .csv is working it doesnt pick .json
about mobile it doesnt detect .json nor .csv

**Reply by HikaruIchijyo** [reply]
also another thing can we automatically detect folder word & add that data to tags? upon import
instead of simple text writter folder & name?

**Reply by HikaruIchijyo** [reply]
> > Can you share the file with only 1 item, no _real_ username or password (change to fake values), so I can investigate further? I can't see how to help any other way, I'm afraid.
> 
> https://file.io/rMuKRpGt73wf
> about web .csv is working it doesnt pick .json
> about mobile it doesnt detect .json nor .csv

simple link above one was to download file https://katb.in/lalafusopum.json

**Reply by HikaruIchijyo** [reply]
oh btw on web only csv is abailable u cant change it to json etc

**Reply by HikaruIchijyo** [reply]
btw csv does work on web & i say it is good i mean new import format

**Reply by BrunoBernardino** [reply]
Hi @HikaruIchijyo , I just tried your sample file (I had to add a final `]}` because it was missing, I assume you manually truncated it too soon), and was able to import the items without any problem:

![Screenshot from 2022-09-12 06-53-27](https://user-images.githubusercontent.com/1239616/189583017-8540658c-a580-4c28-b638-8efa1d729b9c.png)

![Screenshot from 2022-09-12 06-53-43](https://user-images.githubusercontent.com/1239616/189583033-daaf856e-eb27-4a87-beef-6f28a38367c2.png)

It does seem, however, you were able to get it fixed with the CSV format anyway, so I'm happy to close this if you've sorted it.

Let me know!

**Reply by HikaruIchijyo** [reply]
> Hi @HikaruIchijyo , I just tried your sample file (I had to add a final `]}` because it was missing, I assume you manually truncated it too soon), and was able to import the items without any problem:
> 
> ![Screenshot from 2022-09-12 06-53-27](https://user-images.githubusercontent.com/1239616/189583017-8540658c-a580-4c28-b638-8efa1d729b9c.png)
> 
> ![Screenshot from 2022-09-12 06-53-43](https://user-images.githubusercontent.com/1239616/189583033-daaf856e-eb27-4a87-beef-6f28a38367c2.png)
> 
> It does seem, however, you were able to get it fixed with the CSV format anyway, so I'm happy to close this if you've sorted it.
> 
> Let me know!

yeah it was cuz i removed it and forgot to add it back

the only thing left is converting folders to tags idk how to do that

**Reply by BrunoBernardino** [reply]
The import should do that automatically as you can see from the tags above (it adds the file name + folder names as tags).

The fact you can't import with your own file means there should be some parsing error, which can come from some value. Do you have any kind of local JSON file editor (VS Code, for example)? If so, I'd recommend you try verifying/fixing/confirming the file is valid JSON.

**Reply by HikaruIchijyo** [reply]
> The import should do that automatically as you can see from the tags above (it adds the file name + folder names as tags).
> 
> The fact you can't import with your own file means there should be some parsing error, which can come from some value. Do you have any kind of local JSON file editor (VS Code, for example)? If so, I'd recommend you try verifying/fixing/confirming the file is valid JSON.

lemme chk

**Reply by BrunoBernardino** [reply]
Any update on this @HikaruIchijyo ?

**Reply by HikaruIchijyo** [reply]
> Any update on this @HikaruIchijyo ?

Sorry i tried it two to three times previously & it didnt work will chk today again & let u know

**Reply by BrunoBernardino** [reply]
@HikaruIchijyo we've found the issue and merged a fix, but it hasn't been released yet. You can try again on the next release (or, if you're running locally, merge in the latest `main`).

**Reply by HikaruIchijyo** [reply]
> @HikaruIchijyo we've found the issue and merged a fix, but it hasn't been released yet. You can try again on the next release (or, if you're running locally, merge in the latest `main`).

Gotcha will w8 fr next release

**Comment by doublehelix** [comment]
I'm having a similar set of issues...
I have an empty vault, go to "settings > tools > import", select a BitWarden JSON export file.
The file is 862KB, and contains 35 folders (including embedded folders) and around 1800 passwords/notes etc.
The JSON itself is valid.

When I select the file, I get this:
(apparently 0 entries to import)
![image](https://user-images.githubusercontent.com/355130/194789119-3a049932-9a21-416f-8395-5b3b377d2fe3.png)

If I cancel, I cannot retry import without shutting down padloc. If I try and import again, after selecting the file it goes back to the main interface and the import data dialog is not displayed at all.

**Reply by doublehelix** [reply]
Haven't tried to build for debugging this, but my 17th entry (type 1 which is a website password) has no `uris` value (empty array).
If I remove ALL entries AFTER this one, it fails... if I also delete this entry, I am able to successfully import the 16 entries prior to this one.

e.g. (with fake GUID's and name):
```json
        {
            "id": "00000000-0000-0000-0000-000000000000",
            "organizationId": null,
            "folderId": "0000-0000-0000-0000-000000000000",
            "type": 1,
            "reprompt": 0,
            "name": "my website",
            "notes": null,
            "favorite": false,
            "login": {
                "uris": [],
                "username": "username",
                "password": "password",
                "totp": null
            },
            "collectionIds": null
        }
```

**Reply by doublehelix** [reply]
Confirmed that if I change this `"uris": []` value to include an object, I can successfully import 1872 items.

Import Error:
```json
    "uris": []
```

Import Success:
```json
    "uris": [
        {
            "match": null,
            "uri": "http://10.0.0.0"
        }
    ]
```

**Reply by BrunoBernardino** [reply]
Thanks for that helpful tip @doublehelix ! I'll make sure we don't fail parsing on those cases!

**Comment by BrunoBernardino** [comment, marked as answer]
This will be fixed in the next release (4.2.0 or 4.1.1). Thanks for the details!

PR with the fix is at https://github.com/padloc/padloc/pull/617
